Develop profile

yeetivity opened the pull request:
Improved the profile page

- User can now input user specifics
- Specifics are saved to user profile in firebase
- Specifics are read from firestore and displayed in user profile

yeetivity committed:
Created profile layout

yeetivity committed:
Added edit profile activity

yeetivity committed:
Created MVVM structure of user profile

yeetivity committed:
Added editprofile into the MVVM structure

Editprofile can now edit the userprofile in the repo. It is not saved yet to the database.

yeetivity committed:
Added datastorage to firebase

During registering user now automatically saves fullname and email to the firebase database

yeetivity committed:
Saving to firestore is now added to profile as well

In profile activity user can store weight, length, first and last name

yeetivity committed:
Get data from firestore and display it on the profile page.

Still has a bug; when you first open the profile page, the acquiring of data is too slow, which causes the profile page to not be updated. However, an oncreate fixes this. It is somewhere in the MVVM structure.

yeetivity committed:
Made the bug a bit more professional.

yeetivity committed:
BUG FIXED

yeetivity committed:
Added horizontal layouts

elisakaisa commented on app/src/main/java/kth/jjve/xfran/EditProfileActivity.java:
```diff
@@ -0,0 +1,138 @@
+
+        /*-----  VM  -----*/
+        //Todo: check if we can somehow inherit this between editprofile and normal profile
+        UserProfileVM mUserProfileVM = ViewModelProviders.of(this).get(UserProfileVM.class);
+        mUserProfileVM.init();
+        mUserProfileVM.getUserProfile().observe(this, uP -> {
+            if (!uP.checkEmpty()){
+                mFirstName.setText(uP.getFirstName());
+                mLastName.setText(uP.getLastName());
+                mEmail.setText(uP.getEmail());
+                mWeight.setText(Double.toString(uP.getWeight()));
+                mLength.setText(Integer.toString(uP.getLength()));
+            }
+        });
```
Could this be made into a method? I mean it wouldn't change anything, except making the onCreate a bit shorter, which I find less confusing

yeetivity replied:
I think line 52-54 should be in onCreate but line 55-60 could be a method; yes

elisakaisa commented on app/src/main/java/kth/jjve/xfran/EditProfileActivity.java:
```diff
@@ -0,0 +1,138 @@
+        save.setOnClickListener(v -> {
+            // Since a double/int can't be directly read from a edittext,
+            // it is first read as string and then parsed
+            try {
+                String w = mWeight.getText().toString();
+                String l = mLength.getText().toString();
+                double weight = Double.parseDouble(w);
+                int length = Integer.parseInt(l);
+
+                mUserProfileVM.setUserProfile(
+                        mFirstName.getText().toString(),
+                        mLastName.getText().toString(),
+                        mEmail.getText().toString(),
+                        weight,
+                        length);
+
+                Toast.makeText(getApplicationContext(), "User profile saved", Toast.LENGTH_SHORT).show();
+                finish();
+            } catch (Exception e) {
+                e.printStackTrace();
+                Toast.makeText(getApplicationContext(), "Edittexts cannot be empty", Toast.LENGTH_SHORT).show();
+            }
```
Same as above, what if it was simply a method in this class?

yeetivity replied:
Yep

elisakaisa commented on app/src/main/java/kth/jjve/xfran/models/UserProfile.java:
```diff
@@ -0,0 +1,57 @@
+        this.userLastName = lastName;
+        this.userEmail = email;
+        this.userWeight = weight;
+        this.userLength = length;
```
You mean the height of the user and not the length?

elisakaisa commented on app/src/main/res/values/strings.xml:
```diff
@@ -18,4 +18,25 @@
+    <string name="editProfile_firstname_hint">First Name</string>
+    <string name="editProfile_lastname_hint">Last Name</string>
+    <string name="editProfile_weight_hint">Weight [kg]</string>
+    <string name="editProfile_length_hint">Length [cm]</string>
```
Yeah height is defo better

elisakaisa reviewed (approved):
Seems good to me, added a couple of comments on minor details

yeetivity committed:
Processed Elisa's comments

Discovered a new bug (logging out and logging in with other profile kept old profile); also fixed it (userID is now compared to the previous userID in getUserProfile)

yeetivity committed:
Made darktheme a bit more bearable

MariahSabioni commented on app/src/main/java/kth/jjve/xfran/EditProfileActivity.java:
```diff
@@ -0,0 +1,150 @@
+            mLastName.setText(userProfile.getLastName());
+            mEmail.setText(userProfile.getEmail());
+            mWeight.setText(Double.toString(userProfile.getWeight()));
+            mLength.setText(Integer.toString(userProfile.getHeight()));
```
can we change every appearance of 'length' to 'height'?

MariahSabioni commented on app/src/main/java/kth/jjve/xfran/EditProfileActivity.java:
```diff
@@ -0,0 +1,150 @@
+            finish();
+        } catch (Exception e) {
+            e.printStackTrace();
+            Toast.makeText(getApplicationContext(), "Edittexts cannot be empty", Toast.LENGTH_SHORT).show();
```
I think this block has a bug. When I delete my first and last name it doesn't throw an exception and weird things happen - it saves, but it doesn't change on profile activity, but it still appears empty if I return to edit profile activity.

MariahSabioni commented on app/src/main/java/kth/jjve/xfran/ProfileActivity.java:
```diff
@@ -1,14 +1,123 @@
+    @Override
+    protected void onResume() {
+        super.onResume();
+        navigationView.setCheckedItem(R.id.nav_profile);
```
When you update an info on edit profile, it doesnt show the new info on the profile activity after you save it. Thats because it runs on resume and not on create when you "come back" from edit profile. So the updated info will only appear if you change to another activity and come back. Maybe put something on resume to update info (or after saving).

yeetivity replied:
will do later

MariahSabioni reviewed (changes requested):
I still need to understand better how the repo works but otherwise looks awesome. Just requested changes on those 2 minor bugs.

yeetivity committed:
Fixed Mariah Bugs
